modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala

```diff
@@ -209,49 +222,95 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
+          case Left(e) =>
+            if (current.addition === 0)
+              Monad[F].pure(Left(SchemaResolutionError(currentSchemaKey, e)))
+            else if (current.revision < maxSchemaKey.version.revision && isNotFound(e))
+              go(current.copy(revision = current.revision + 1, addition = 0), acc)
+            else
+              Monad[F].pure(Left(SchemaResolutionError(currentSchemaKey, e)))
+          case Right(json) =>
+            if (current.revision < maxSchemaKey.version.revision)
+              go(
+                current.copy(addition = current.addition + 1),
+                SelfDescribingSchema(SchemaMap(currentSchemaKey), json) :: acc
               )
-            )
+            else if (current.addition < maxSchemaKey.version.addition)
+              go(
+                current.copy(addition = current.addition + 1),
+                SelfDescribingSchema(SchemaMap(currentSchemaKey), json) :: acc
+              )
+            else
+              Monad[F].pure(
+                Right(
+                  NonEmptyList(SelfDescribingSchema(SchemaMap(currentSchemaKey), json), acc).reverse
+                )
+              )
+        }
       }
+
+      go(SchemaVer.Full(maxSchemaKey.version.model, 0, 0), Nil)
     }
 
-    go(SchemaVer.Full(maxSchemaKey.version.model, 0, 0), Nil)
+    def handleAfterFetch(
+      result: Either[SchemaResolutionError, SchemaContentList]
+    ): F[Either[SchemaResolutionError, SchemaContentListLookupResult]] =
+      cache match {
+        case Some(c) =>
+          val updated = result.leftMap(e => resolutionErrorToFailureMap(e))
+          c.putSchemaContentListResult(maxSchemaKey, updated).map {
+            case Right(ResolverCache.TimestampedItem(i, t)) =>
+              Right(ResolverResult.Cached(maxSchemaKey, i, t))
+            case Left(failure) =>
+              val schemaKey = result.leftMap(_.schemaKey).left.getOrElse(maxSchemaKey)
+              Left(SchemaResolutionError(schemaKey, resolutionError(failure)))
+          }
+        case None =>
+          result
+            .map[SchemaContentListLookupResult](ResolverResult.NotCached(_))
+            .pure[F]
+      }
+
+    def lockAndLookup: F[Either[SchemaResolutionError, SchemaContentListLookupResult]] =
+      withLockOnSchemaContentList(maxSchemaKey) {
+        getSchemaContentListFromCache(maxSchemaKey).flatMap {
+          case Some(TimestampedItem(Right(i), t)) =>
+            Monad[F].pure(Right(ResolverResult.Cached(maxSchemaKey, i, t)))
+          case Some(TimestampedItem(Left(_), _)) | None =>
+            for {
+              result <- get()
+              fixed  <- handleAfterFetch(result)
+            } yield fixed
+        }
+      }
+
+    getSchemaContentListFromCache(maxSchemaKey).flatMap {
+      case Some(TimestampedItem(Right(i), t)) =>
+        Monad[F].pure(Right(ResolverResult.Cached(maxSchemaKey, i, t)))
+      case Some(TimestampedItem(Left(_), _)) | None =>
+        lockAndLookup
+    }
```

Comment:
On line 305 we know that the first call to `getSchemaContentListFromCache` returned either `Left` or `None`.
Yet on line 290 after we call it a second time we check again if it is successful or not.
Is it because there might be several parallel calls to the function and the cache can get updated between the 2 calls ?

Reply:
Exactly!